python_modules/dagster/dagster/core/log.py

```diff
@@ -80,14 +126,46 @@ def _log(self, method, orig_message, message_props):
         #     message_with_structured_props, extra={DAGSTER_META_KEY: all_props}
         # )
 
-        for logger in self.loggers:
-            logger_method = check.is_callable(getattr(logger, method))
-            if logger.name == DAGSTER_DEFAULT_LOGGER:
-                logger_method(
-                    msg_with_multiline_structured_props, extra={DAGSTER_META_KEY: all_props}
-                )
+        # HACK: getLevelName really doesn't do what it says on the box. This function was originally
+        # intended to give the user-facing name of a given int log level, e.g.,
+        # getLevelName(40) -> 'ERROR'. In fact, it also works the other way around (this was 'fixed'
+        # in Python 3.4 then immediately reverted in 3.4.2). For levels it doesn't recognize,
+        # including all ints that don't have a name registered with logging.addLevelName it does:
+        # getLevelName('foo') -> 'Level foo'; getLevelName(3) -> 'Level 3' (!)
+        # See: https://docs.python.org/3/library/logging.html#logging.getLevelName
+        if isinstance(level, int):
+            lvl = level
+        else:
+            lvl = logging.getLevelName(level.upper())
+
+        if not isinstance(lvl, int):
+            self.error(
+                'Unexpected log level: User code attempted to log at level \'{level}\', but that '
+                'level has not been registered with the Python logging library. Original message: '
+                '\'{orig_message}\''.format(level=level, orig_message=orig_message),
+                **message_props
+            )
+            return
+
+        for logger_ in self.loggers:
+            if not isinstance(level, int) and hasattr(logger_, level):
+                logger_method = check.is_callable(getattr(logger_, level))
+                if logger_.name == DAGSTER_DEFAULT_LOGGER:
```

Comment:
i'm a bit confused where there is a specialized behavior for the default logger

Reply:
The multiline thing this seems like it should a property of the logger rather than hardcoded. A specialized "formatter" or something.

Reply:
++

Reply:
this is just a holdover from the status quo ante, I think from a transitional state in multiline logging where dagit was still relying on single-line logging to function.

Reply:
I actually don't think there's any reason not to enforce multiline formatting everywhere anymore